File: Toggly.FeatureManagement.iOS/TogglySwiftUI/Sources/FeatureFlag.swift
import SwiftUI
import TogglyCore

/// A property wrapper that provides reactive access to a feature flag.
///
/// Usage:
/// ```swift
/// struct MyView: View {
///     @FeatureFlag("new-feature") var isNewFeatureEnabled
///
///     var body: some View {
///         if isNewFeatureEnabled {
///             NewFeatureView()
///         } else {
///             LegacyView()
///         }
///     }
/// }
/// ```
@propertyWrapper
public struct FeatureFlag: DynamicProperty {
    @StateObject private var observer: FeatureFlagObserver

    /// The current value of the feature flag.
    public var wrappedValue: Bool {
        observer.isEnabled
    }

    /// A binding to the feature flag value.
    public var projectedValue: Binding<Bool> {
        Binding(
            get: { observer.isEnabled },
            set: { _ in } // Read-only
        )
    }

    /// Creates a feature flag property wrapper.
    /// - Parameters:
    ///   - key: The feature flag key.
    ///   - defaultValue: The default value if the flag is not found.
    ///   - negate: Whether to invert the result (preferred off-path check).
    ///   - context: Optional per-evaluation entity context.
    ///   - kind: Optional kind for `registerContext` mapper lookup.
    ///   - service: The Toggly service to use. Defaults to the shared instance.
    public init(
        _ key: String,
        defaultValue: Bool = false,
        negate: Bool = false,
        context: Any? = nil,
        kind: String? = nil,
        service: TogglyService? = nil
    ) {
        _observer = StateObject(wrappedValue: FeatureFlagObserver(
            key: key,
            defaultValue: defaultValue,
            negate: negate,
            context: context,
            kind: kind,
            service: service
        ))
    }
}

/// An observable object that tracks a feature flag's state.
@MainActor
public final class FeatureFlagObserver: ObservableObject {
    /// Whether the feature check currently passes (after optional negate).
    @Published public private(set) var isEnabled: Bool

    private let key: String
    private let defaultValue: Bool
    private let negate: Bool
    private let context: Any?
    private let kind: String?
    private let service: TogglyService?
    private var unsubscribe: (@Sendable () -> Void)?

    init(
        key: String,
        defaultValue: Bool,
        negate: Bool = false,
        context: Any? = nil,
        kind: String? = nil,
        service: TogglyService?
    ) {
        self.key = key
        self.defaultValue = defaultValue
        self.negate = negate
        self.context = context
        self.kind = kind
        self.service = service
        self.isEnabled = negate ? !defaultValue : defaultValue

        Task {
            await self.setup()
        }
    }

    private func setup() async {
        let toggly = service ?? (Toggly.isConfigured ? Toggly.shared : nil)

        guard let toggly = toggly else {
            isEnabled = negate ? !defaultValue : defaultValue
            return
        }

        isEnabled = await evaluate(with: toggly)

        unsubscribe = await toggly.addStateChangeHandler { [weak self] featureKey, _, _ in
            guard let self = self, featureKey == self.key else { return }
            Task { @MainActor in
                guard let toggly = self.service ?? (Toggly.isConfigured ? Toggly.shared : nil) else { return }
                self.isEnabled = await self.evaluate(with: toggly)
            }
        }
    }

    private func evaluate(with toggly: TogglyService) async -> Bool {
        await toggly.evaluateFeatureGate(
            featureKeys: [key],
            requirement: .all,
            negate: negate,
            context: context,
            kind: kind
        )
    }

    deinit {
        unsubscribe?()
    }
}

/// A property wrapper for multiple feature flags with requirement logic.
@propertyWrapper
public struct FeatureGate: DynamicProperty {
    @StateObject private var observer: FeatureGateObserver

    /// The current value of the feature gate.
    public var wrappedValue: Bool {
        observer.isEnabled
    }

    /// Creates a feature gate property wrapper.
    /// - Parameters:
    ///   - keys: The feature flag keys to evaluate.
    ///   - requirement: Whether all or any flags must be enabled.
    ///   - negate: Whether to negate the result.
    ///   - context: Optional per-evaluation entity context.
    ///   - kind: Optional kind for `registerContext` mapper lookup.
    ///   - service: The Toggly service to use.
    public init(
        _ keys: [String],
        requirement: FeatureRequirement = .all,
        negate: Bool = false,
        context: Any? = nil,
        kind: String? = nil,
        service: TogglyService? = nil
    ) {
        _observer = StateObject(wrappedValue: FeatureGateObserver(
            keys: keys,
            requirement: requirement,
            negate: negate,
            context: context,
            kind: kind,
            service: service
        ))
    }
}

/// An observable object that tracks a feature gate's state.
@MainActor
public final class FeatureGateObserver: ObservableObject {
    @Published public private(set) var isEnabled: Bool = false

    private let keys: [String]
    private let requirement: FeatureRequirement
    private let negate: Bool
    private let context: Any?
    private let kind: String?
    private let service: TogglyService?
    private var unsubscribe: (@Sendable () -> Void)?

    init(
        keys: [String],
        requirement: FeatureRequirement,
        negate: Bool,
        context: Any? = nil,
        kind: String? = nil,
        service: TogglyService?
    ) {
        self.keys = keys
        self.requirement = requirement
        self.negate = negate
        self.context = context
        self.kind = kind
        self.service = service

        Task {
            await self.setup()
        }
    }

    private func setup() async {
        let toggly = service ?? (Toggly.isConfigured ? Toggly.shared : nil)

        guard let toggly = toggly else {
            isEnabled = negate // Default to false, negated if needed
            return
        }

        isEnabled = await evaluate(with: toggly)

        unsubscribe = await toggly.addStateChangeHandler { [weak self] featureKey, _, _ in
            guard let self = self, self.keys.contains(featureKey) else { return }
            Task { @MainActor in
                guard let toggly = self.service ?? (Toggly.isConfigured ? Toggly.shared : nil) else { return }
                self.isEnabled = await self.evaluate(with: toggly)
            }
        }
    }

    private func evaluate(with toggly: TogglyService) async -> Bool {
        await toggly.evaluateFeatureGate(
            featureKeys: keys,
            requirement: requirement,
            negate: negate,
            context: context,
            kind: kind
        )
    }

    deinit {
        unsubscribe?()
    }
}

File: Toggly.FeatureManagement.Android/toggly-compose/src/main/kotlin/io/toggly/compose/FeatureFlag.kt
package io.toggly.compose

import androidx.compose.runtime.*
import io.toggly.core.models.FeatureRequirement

/**
 * Remember whether a feature is enabled (optionally negated / entity-scoped).
 *
 * Prefer [Feature] with `negate = true` for the off path instead of a separate Off API.
 *
 * @param featureKey The feature key to check
 * @param negate Whether to invert the result (show when the feature is off)
 * @param context Optional per-evaluation entity context
 * @param contextKind Optional kind for [io.toggly.core.registerContext] mapper lookup
 * @param defaultValue Default value while loading / before first evaluation
 * @return Whether content should show for this feature check
 */
@Composable
fun rememberFeature(
    featureKey: String,
    negate: Boolean = false,
    context: Any? = null,
    contextKind: String? = null,
    defaultValue: Boolean = false
): Boolean {
    return rememberFeatureGate(
        featureKeys = listOf(featureKey),
        requirement = FeatureRequirement.ALL,
        negate = negate,
        context = context,
        contextKind = contextKind,
        defaultValue = defaultValue
    )
}

/**
 * Remember whether a feature flag is enabled.
 *
 * @param featureKey The feature key to check
 * @param defaultValue Default value while loading
 * @return Whether the feature is enabled
 */
@Composable
fun rememberFeatureFlag(
    featureKey: String,
    defaultValue: Boolean = false
): Boolean = rememberFeature(featureKey = featureKey, defaultValue = defaultValue)

/**
 * Remember whether a feature flag is enabled, returning a State.
 *
 * @param featureKey The feature key to check
 * @param defaultValue Default value while loading
 * @return State containing whether the feature is enabled
 */
@Composable
fun rememberFeatureFlagAsState(
    featureKey: String,
    defaultValue: Boolean = false
): State<Boolean> {
    val service = LocalTogglyService.current

    return service.featureFlagFlow(featureKey)
        .collectAsState(initial = defaultValue)
}

/**
 * Remember the result of a feature gate evaluation.
 *
 * When [context] is null, evaluates against the boolean snapshot (entity gates → false).
 * When [context] is provided, uses the service's entity-aware evaluation.
 *
 * @param featureKeys The feature keys to evaluate
 * @param requirement Whether all or any features must be enabled
 * @param negate Whether to negate the result
 * @param context Optional per-evaluation entity context
 * @param contextKind Optional kind for registerContext mapper lookup
 * @param defaultValue Default value while loading
 * @return Whether the gate passes
 */
@Composable
fun rememberFeatureGate(
    featureKeys: List<String>,
    requirement: FeatureRequirement = FeatureRequirement.ALL,
    negate: Boolean = false,
    context: Any? = null,
    contextKind: String? = null,
    defaultValue: Boolean = false
): Boolean {
    val service = LocalTogglyService.current
    val featureFlags = LocalFeatureFlags.current

    if (context == null && contextKind == null) {
        return remember(featureFlags, featureKeys, requirement, negate) {
            evaluateSnapshotFeatureGate(featureFlags, featureKeys, requirement, negate)
        }
    }

    return produceState(
        initialValue = defaultValue,
        service,
        featureFlags,
        featureKeys,
        requirement,
        negate,
        context,
        contextKind
    ) {
        value = service.evaluateFeatureGate(
            featureKeys,
            requirement,
            negate,
            context,
            contextKind
        )
    }.value
}

/**
 * Remember the result of a feature gate evaluation as State.
 *
 * @param featureKeys The feature keys to evaluate
 * @param requirement Whether all or any features must be enabled
 * @param negate Whether to negate the result
 * @param defaultValue Default value while loading
 * @return State containing whether the gate passes
 */
@Composable
fun rememberFeatureGateAsState(
    featureKeys: List<String>,
    requirement: FeatureRequirement = FeatureRequirement.ALL,
    negate: Boolean = false,
    defaultValue: Boolean = false
): State<Boolean> {
    val service = LocalTogglyService.current

    return service.featureGateFlow(featureKeys, requirement, negate)
        .collectAsState(initial = defaultValue)
}

/**
 * Composable that renders [content] when a feature check passes.
 *
 * Use `negate = true` for the off path (preferred over [FeatureFlagOff]).
 *
 * @param featureKey The feature key to check
 * @param negate Invert the check (show content when the feature is off)
 * @param context Optional per-evaluation entity context
 * @param contextKind Optional kind for registerContext mapper lookup
 * @param fallback Optional fallback when the check fails
 * @param content Content to show when the check passes
 */
@Composable
fun Feature(
    featureKey: String,
    negate: Boolean = false,
    context: Any? = null,
    contextKind: String? = null,
    fallback: @Composable (() -> Unit)? = null,
    content: @Composable () -> Unit
) {
    val shouldShow = rememberFeature(
        featureKey = featureKey,
        negate = negate,
        context = context,
        contextKind = contextKind
    )

    if (shouldShow) {
        content()
    } else {
        fallback?.invoke()
    }
}

/**
 * Composable that only renders content if a feature is enabled.
 *
 * Prefer [Feature]. Kept for source compatibility.
 *
 * @param featureKey The feature key to check
 * @param fallback Optional fallback content when feature is disabled
 * @param content Content to show when feature is enabled
 */
@Deprecated(
    message = "Use Feature(featureKey, ...) instead",
    replaceWith = ReplaceWith(
        "Feature(featureKey = featureKey, fallback = fallback, content = content)",
        "io.toggly.compose.Feature"
    )
)
@Composable
fun FeatureFlag(
    featureKey: String,
    fallback: @Composable (() -> Unit)? = null,
    content: @Composable () -> Unit
) {
    Feature(
        featureKey = featureKey,
        fallback = fallback,
        content = content
    )
}

/**
 * Composable that only renders content if a feature is disabled.
 *
 * Prefer [Feature] with `negate = true`.
 *
 * @param featureKey The feature key to check
 * @param fallback Optional fallback content when feature is enabled
 * @param content Content to show when feature is disabled
 */
@Deprecated(
    message = "Use Feature(featureKey, negate = true) instead",
    replaceWith = ReplaceWith(
        "Feature(featureKey = featureKey, negate = true, fallback = fallback, content = content)",
        "io.toggly.compose.Feature"
    )
)
@Composable
fun FeatureFlagOff(
    featureKey: String,
    fallback: @Composable (() -> Unit)? = null,
    content: @Composable () -> Unit
) {
    Feature(
        featureKey = featureKey,
        negate = true,
        fallback = fallback,
        content = content
    )
}

/**
 * Composable that renders content based on a feature gate evaluation.
 *
 * @param featureKeys The feature keys to evaluate
 * @param requirement Whether all or any features must be enabled
 * @param negate Whether to negate the result
 * @param context Optional per-evaluation entity context
 * @param contextKind Optional kind for registerContext mapper lookup
 * @param fallback Optional fallback content when gate fails
 * @param content Content to show when gate passes
 */
@Composable
fun FeatureGate(
    featureKeys: List<String>,
    requirement: FeatureRequirement = FeatureRequirement.ALL,
    negate: Boolean = false,
    context: Any? = null,
    contextKind: String? = null,
    fallback: @Composable (() -> Unit)? = null,
    content: @Composable () -> Unit
) {
    val isEnabled = rememberFeatureGate(
        featureKeys = featureKeys,
        requirement = requirement,
        negate = negate,
        context = context,
        contextKind = contextKind
    )

    if (isEnabled) {
        content()
    } else {
        fallback?.invoke()
    }
}

/**
 * Variant-style dual-slot composable (on vs off content in one call).
 *
 * Not the primary off API — prefer two [Feature] calls or [Feature] with `negate = true`
 * for the off path. Kept for dual-slot layouts.
 *
 * @param featureKey The feature key to check
 * @param enabled Content to show when feature is enabled
 * @param disabled Content to show when feature is disabled
 */
@Composable
fun FeatureSwitch(
    featureKey: String,
    enabled: @Composable () -> Unit,
    disabled: @Composable () -> Unit
) {
    val isEnabled = rememberFeature(featureKey)

    if (isEnabled) {
        enabled()
    } else {
        disabled()
    }
}
